src/Symfony/Component/Routing/CompiledRoute.php

```diff
@@ -63,14 +63,6 @@ public function __serialize(): array
         ];
     }
 
-    /**
-     * @internal since Symfony 4.3, will be removed in Symfony 5 as the class won't implement Serializable anymore
-     */
-    public function serialize()
```

Comment:
@nicolas-grekas removing this but keeping the new magic serialization methods means that the deserialization breaks when upgrading php to 7.4. so persisted routes will not work anymore when upgrading php. I think that was one of the main reasons why we implemented Serializable in the first place.

Reply:
So, we can keep them for a smoother transition, and keep only `@internal` then, would that be ok to you?

Reply:
Oh, let's make them really `final` to close this a bit more?

Reply:
If we make them final Drupal can't overwrite them anymore which is what they do in https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Routing%21CompiledRoute.php/class/CompiledRoute/8.2.x 
I don't see what else they could do then. It would make it impossible to adjust serialization of in a subclass.

Reply:
@nikic In https://wiki.php.net/rfc/custom_object_serialization you say

> For this reason, code using parent::serialize() is generally broken

But from what I see the implementation in Drupal calling parent::serialize does not seem broken

```php
public function serialize() {
    $data = unserialize(parent::serialize());
    $data['fit'] = $this->fit;
    $data['patternOutline'] = $this->patternOutline;
    $data['numParts'] = $this->numParts;
    return serialize($data);
  }
  public function unserialize($serialized) {
    parent::unserialize($serialized);
    $data = unserialize($serialized);
    $this->fit = $data['fit'];
    $this->patternOutline = $data['patternOutline'];
    $this->numParts = $data['numParts'];
  }
```

It's not using backreferences but if it was, this approach using `unserialize(parent::serialize()` seems fine.

Reply:
I would propose to revert #30286 for the routing component.

Reply:
We provide `__serialize` and `__unserialize` as extension points. Note that yes, routes are exposed to the issue. The reason is that options/defaults can be anything, like objects with these references that break. Drupal is doing that (objects there).

This means to me we should keep implementing Serializable, but make the corresponding method `@internal` - and tell ppl to use `__serialize` and `__unserialize` as extension points instead.

This will provide FC/BC.